File: TasksList/CSharp/Web/AspNetMinimalApi/src/Services/TaskService.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;

public class TaskService
{
    private readonly TasksListDbContext _db;
    private readonly ICacheAdapter _cache;

    public TaskService(TasksListDbContext db, ICacheAdapter cache)
    {
        _db = db;
        _cache = cache;
    }

    public List<TaskItem> GetAll()
    {
        var cached = _cache.Get("tasks:all");
        if (cached != null)
            return JsonSerializer.Deserialize<List<TaskItem>>(cached)!;
        var list = _db.Tasks.OrderBy(t => t.Id).ToList();
        _cache.Set("tasks:all", JsonSerializer.Serialize(list));
        return list;
    }

    public TaskItem Create(string title, string description)
    {
        var task = new TaskItem { Title = title, Description = description, Completed = false, CreatedAt = DateTime.UtcNow.ToString("o") };
        _db.Tasks.Add(task);
        _db.SaveChanges();
        _cache.Delete("tasks:all");
        return task;
    }

    public bool Complete(int id)
    {
        var task = _db.Tasks.Find(id);
        if (task == null) return false;
        task.Completed = true;
        _db.Tasks.Update(task);
        _db.SaveChanges();
        _cache.Delete("tasks:all");
        return true;
    }

    public bool Delete(int id)
    {
        var task = _db.Tasks.Find(id);
        if (task == null) return false;
        _db.Tasks.Remove(task);
        _db.SaveChanges();
        _cache.Delete("tasks:all");
        return true;
    }
}

File: TasksList/Java/Web/SpringBoot/src/main/java/com/portfolio/tasks/TaskService.java
package com.portfolio.tasks;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class TaskService {
    private final TaskRepository repository;
    private final CacheAdapter cache;
    private final ObjectMapper objectMapper;

    public TaskService(TaskRepository repository, CacheAdapter cache, ObjectMapper objectMapper) {
        this.repository = repository;
        this.cache = cache;
        this.objectMapper = objectMapper;
    }

    public List<Task> getAll() {
        String cached = cache.get("tasks:all");
        if (cached != null) {
            return deserializeList(cached);
        }
        List<Task> list = repository.findAll().stream()
                .sorted((a, b) -> Long.compare(a.getId(), b.getId()))
                .toList();
        cache.set("tasks:all", serialize(list));
        return list;
    }

    public Optional<Task> getById(Long id) {
        return repository.findById(id);
    }

    public Task create(String title) {
        Task task = repository.save(new Task(title));
        cache.delete("tasks:all");
        return task;
    }

    public Optional<Task> update(Long id, String title, Boolean completed) {
        Optional<Task> existing = repository.findById(id);
        if (existing.isEmpty()) {
            return Optional.empty();
        }
        Task task = existing.get();
        if (title != null) {
            task.setTitle(title);
        }
        if (completed != null) {
            task.setCompleted(completed);
        }
        repository.save(task);
        cache.delete("tasks:all");
        return Optional.of(task);
    }

    public boolean delete(Long id) {
        Optional<Task> task = repository.findById(id);
        if (task.isEmpty()) {
            return false;
        }
        repository.delete(task.get());
        cache.delete("tasks:all");
        return true;
    }

    private String serialize(List<Task> list) {
        try {
            return objectMapper.writeValueAsString(list);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    private List<Task> deserializeList(String json) {
        try {
            return objectMapper.readValue(json, new TypeReference<>() {
            });
        } catch (Exception e) {
            return List.of();
        }
    }
}
